platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/ProfileMigratable.java

```diff
@@ -0,0 +1,201 @@
+    Validate.notNull(featureMigrator, "invalid null feature migrator");
+    Validate.notNull(bundleMigrator, "invalid null bundle migrator");
+    this.appMigrator = appMigrator;
+    this.featureMigrator = featureMigrator;
+    this.bundleMigrator = bundleMigrator;
+  }
+
+  @Override
+  public String getVersion() {
+    return ProfileMigratable.VERSION;
+  }
+
+  @Override
+  public String getId() {
+    return "ddf.profile";
+  }
+
+  @Override
+  public String getTitle() {
+    return "Profile Migratable";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Exports the state of applications, features, and bundles";
+  }
+
+  @Override
+  public String getOrganization() {
+    return "Codice";
+  }
+
+  @Override
+  public void doExport(ExportMigrationContext context) {
+    context
+        .getEntry(ProfileMigratable.PROFILE_PATH)
+        .store(
+            (report, os) ->
+                JsonUtils.writeValue(
+                    os,
+                    new JsonProfile(
+                        appMigrator.exportApplications(),
+                        featureMigrator.exportFeatures(),
+                        bundleMigrator.exportBundles())));
+  }
+
+  @Override
+  public void doImport(ImportMigrationContext context) {
+    try {
+      context.getEntry(ProfileMigratable.PROFILE_PATH).restore(this::restore);
+    } catch (MigrationException e) { // don't want to abort the import; just record the error
+      context.getReport().record(e);
+    }
+  }
+
+  private boolean restore(MigrationReport report, Optional<InputStream> ois) throws IOException {
+    final InputStream is = ois.orElse(null);
+
+    if (is == null) {
+      throw new MigrationException("Import error: missing exported profile information.");
+    }
+    final JsonProfile jprofile =
+        JsonUtils.fromJson(IOUtils.toString(is, StandardCharsets.UTF_8), JsonProfile.class);
+
+    for (int i = 1; i < ProfileMigratable.RETRY_COUNT; i++) {
+      LOGGER.debug("importing system profile ({} out of {})", i, ProfileMigratable.RETRY_COUNT);
+      final Boolean result = restore(report, jprofile, false);
+
+      if (result != null) {
+        return result;
+      }
+    }
+    LOGGER.debug(
+        "verifying system profile", ProfileMigratable.RETRY_COUNT, ProfileMigratable.RETRY_COUNT);
+    final Boolean result = restore(report, jprofile, true);
+
+    if (result != null) {
+      return result;
+    }
+    // if we get here then we had more tasks to execute after having tried so many times!!!
+    throw new IllegalStateException("too many attempts to import profile");
+  }
+
+  /**
+   * Performs a restore attempt.
+   *
+   * @param report the report where to record errors
+   * @param jprofile the exported profile
+   * @param finalAttempt whether this is a final attempt in which case we no longer suppress errors
+   * @return <code>true</code> if the restore was successfull and didn't need to perform any
```

Comment:
I am not a fan of using such ternary value but I am a lot less of a fan to use an illegal state exception as control branching especially since it is a runtime exception that could popup for other reasons by methods I invoke that are not under my control.

I also did not want to introduce an enum with 3 values.